keps/sig-api-machinery/5051-server-side-apply-unsetting-fields/README.md

```diff
@@ -0,0 +1,988 @@
+
+_none_
+
+## Proposal
+
+Introduce a "marker" value that may be used in apply configurations to indicate that a value should
+be unset.
+
+The proposed marker value is: `{k8s_io__value: unset}`
+
+For example, given an object with a field owned by field manager "mgr1":
+
+```yaml
+apiVersion: stable.example.com/v1
+kind: Example
+metadata:
+  name: example1
+  managedFields:
+  - manager: mgr1
+    operation: Apply
+    …
+    fieldsV1:
+      f:spec:
+        f:field {}
+spec:
+  field: "xyz"
+```
+
+Field manager "mgr2" may unset the field by force applying a apply configuration like:
+
+```yaml
+apiVersion: stable.example.com/v1
+kind: Example
+metadata:
+  name: example1
+spec:
+  field: {k8s_io__value: unset}
+```
+
+After the configuration is applied field will be unset and will be owned by "mgr2".
+
+```yaml
+apiVersion: stable.example.com/v1
+kind: Example
+metadata:
+  name: example1
+  managedFields:
+  - manager: mgr2
+    operation: Apply
+    …
+    fieldsV1:
+      f:spec:
+        f:field {} # mgr2 owns this field even though it is unset
+spec:
+  # empty
+```
+
+### User Stories
+
+#### Apply (force=true) to unset an owned listType=map element
+
+Existing value:
+
+```yaml
+field: [
+  {name: "a", value: 1}, 
+  {name: "b", value: 2}
+]
+```
+
+Existing field management:
+
+```yaml
+fieldManager1: 
+  spec.field[name=a]
+  spec.field[name=b]
+```
+
+Force apply configuration:
+
+```yaml
+field: [
+  {name: "b", k8s_io__value: unset}
+]
+```
+
+Result value:
+
+```yaml
+field: [
+  {name: "a", value: 1}
+]
+```
+
+Result field management:
+
+```yaml
+fieldManager1: 
+  spec.field[name=a]
+
+fieldManager2: 
+  spec.field[name=b]
+```
+
+#### Apply (force=true) to unset an owned granular map value
+
+Existing value:
+
+```yaml
+field: {"a": 1, "b": 2}
+```
+
+Existing field management:
+
+```yaml
+fieldManager1: 
+  spec.field
+```
+
+Force apply configuration:
+
+```yaml
+field: {"b": {k8s_io__value: unset}}
+```
+
+Result value:
+
+```yaml
+field: {"a": 1}
+```
+
+Result field management:
+
+```yaml
+fieldManager2:
+  field
+```
+
+#### Apply (force=false) to an owned field
+
+Existing value:
+
+```yaml
+spec:
+  field: 1
+```
+
+Existing field management:
+
+```yaml
+fieldManager1: 
+  spec.field
+```
+
+Apply configuration:
+
+```yaml
+spec:
+  field: {k8s_io__value: unset}
+```
+
+Apply conflicts:
+
+```yaml
+field owned by fieldManager1
+```
+
+#### Apply (force=false) to a unset field that is owned
+
+Existing value:
+
+```yaml
+spec:
+  # empty
+```
+
+Existing field management:
+
+```yaml
+fieldManager1: 
+  spec.field
+```
+
+Apply configuration:
+
+```yaml
+spec:
+  field: "xyz"
+```
+
+Apply conflicts:
+
+```yaml
+field owned by fieldManager1
+```
+
+#### Apply (force=false) to unset a field that is owned but already unset
+
+Existing value:
+
+```yaml
+spec:
+```
+
+Existing field management:
+
+```yaml
+fieldManager1: 
+  spec.field
+```
+
+Apply configuration:
+
+```yaml
+spec:
+  field: {k8s_io__value: unset}
```

Comment:
when encoded in JSON or CBOR, this will be a string containing this?  Actually what format is this?  it's not quite JSON.

Reply:
This looks like YAML to me. I think the JSON would be something like:
```yaml
  "spec": {
    "field": {
      "k8s_io__value": "unset"
    },
```

Reply:
That's right.  I could switch this example to be more idiomatic YAML if that helps:

```yaml
spec:
  field:
    k8s_io__value: unset
```

Reply:
> This looks like YAML to me. I think the JSON would be something like:

Oh, my eye didn't read it that way.  That means that it is logically impossible to escape the value for a  piece of say integer configuration exposed from one resource to set a value in another.  For instance, a controller of deployments would be unable to express "must be empty" in its API.

That seems really problematic.